mount/mount.go

```diff
@@ -38,3 +45,39 @@ func All(mounts []Mount, target string) error {
 	}
 	return nil
 }
+
+// UnmountAllRecursive unmounts all the provided mounts under a target, starting
+// with the deepest mount first.
+func UnmountAllRecursive(mounts []Mount, target string, flags int) error {
+	var mountpoints []string
+	for _, m := range mounts {
+		mountpoints = append(mountpoints, filepath.Join(target, m.Target))
+	}
+	return unmountRecursive(mountpoints, flags)
+}
+
+func unmountRecursive(mountpoints []string, flags int) error {
+	targetSet := make(map[string]struct{})
+	for _, m := range mountpoints {
+		targetSet[m] = struct{}{}
+	}
+
+	var targets []string
+	for m := range targetSet {
+		targets = append(targets, m)
+	}
+
+	// Make the deepest mount be first
+	sort.Slice(targets, func(i, j int) bool {
+		return len(targets[i]) > len(targets[j])
```

Comment:
Not sure you could safely assume this is the unmount order, perhaps using a prefix check rather than len check.

Reply:
There are some places which uses len check as well:
https://github.com/containerd/containerd/blob/v1.7.0-beta.1/mount/temp_unix.go#L48-L51
https://github.com/containerd/containerd/blob/v1.7.0-beta.1/pkg/cri/server/helpers_linux.go#L170-L173

Giving this a second thought... if there is an unmount order that is invalid, that would mean a parent mount was unmounted prior to a submount. By virtue of being a parent mount, its mountpoint is a strict prefix of the mountpoint of the submount, so a len check will be sufficient in this case. If it isn't a strict prefix, then the unmount order shouldn't matter.